### PR TITLE
Timer&health

### LINKED ISSUE #11: Player Health Functionality

### LINKED ISSUE #12: Player Health UI Spawning

### LINKED ISSUE #83: Rail switching ( side to side rails)

### LINKED ISSUE #84: Jumps out of the rail when end is reached

### LINKED ISSUE #92: Timer & Ui

### PR DESCRIPTION
**Pull request by avinash751**
### Tasks Completed
1. closes #11 
2. closes #12 
3. closes #92
4. closes #83 
5. closes #84 
6. new way to  get on rails ( trying to account player direction)
7. Enemy Health  Functionality 
8. Bit Of Turret Enemy Re structure, mainly about transferring projectile data to instantiated bullet 
9. Damage Feedback, same From dreadrun, though i have cleaned it up a bit, made the code more dry, removed a lot of nesting.

### Entering Rails 
There was never a proper way to enter rails and accounting player direction, this time i tried a different approach by using the dot product of the player and direction of the spline. though this works only 60% of the time. I have spent 4 hrs on trying a solution so far i have no idea what is the solution to this. I will try again next sprint when i am doing a refactor for the whole movement system.

### Timer & UI 
i made a timer utilities class,  that has a bunch of premade variation of different timers, Was it really required to be made, No, but i  really just wanted to make a generic timer utilities class, that can be reused multiple times in  the long long term. because we were constantly making timer scripts in dreadrun. So even if we dont need it,  I maybe will use it in  future projects.

### CODE REVIEW

File: Assets/PROJECT/Scripts/PlayerRelated/PlayerCombat/PlayerHealth.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    public static Action<int,GameObject> OnPlayerHealthInitialized;
+
+    protected override void OnEnable()
+    {
+        base.OnEnable();
+    }
+
+    private void Start()
+    {
+        OnPlayerHealthInitialized?.Invoke(MaxHealth, gameObject);
```
Comment (HassanAbusalih) [inline Assets/PROJECT/Scripts/PlayerRelated/PlayerCombat/PlayerHealth.cs:16]: What is this for?

**Comment by HassanAbusalih** [comment]
I know I'm doing this in the morning, but this is such a headache to look through man. Make two pull requests next time :(

**Comment by Armaantakyar1** [comment]
from a code perspective, it looks good, but #84 the jump doesn't feel good. halfway through the jump the player just stops and gets dragged down lmao
I've added a doddle to explain it better
jump

**Review by Armaantakyar1 — APPROVED** [review]

**Review by devTrivedi1 — APPROVED** [review]
I like the jump direction on how it goes, but it would be nice if you could also be able to add a bit more speed as you jump Im assuming that's still in the works tho

are you taking about the jump when you are on rails ?

**Review by HassanAbusalih — APPROVED** [review]
Shaping up well, definitely need to have dynamic button association for the rails depending on the camera though, the controls feel super weird.

**Comment by avinash751** [comment]
> from a code perspective, it looks good, but #84 the jump after reaching of the end of the rail if you aren't pressing anything doesn't feel good. halfway through the jump the player just stops and gets dragged down lmao I've added a doddle to explain it better jump
ohhh phew for a sec i am like jump does not feel good i was like HUH?, but yeah i get what you mean, ayyt will do in next sprint

File: Assets/PROJECT/Scripts/PlayerRelated/PlayerCombat/PlayerHealth.cs (hunk 1 above)
Comment (avinash751) [reply Assets/PROJECT/Scripts/PlayerRelated/PlayerCombat/PlayerHealth.cs:16]: this for updating health ui , i did  not want to do a hard code reference , thats why?

**Comment by avinash751** [comment]
> I know I'm doing this in the morning, but this is such a headache to look through man. Make two pull requests next time :(

understandable i was thinking whether i do two branches or not then i remembered the merge meeting we had in dreadrun , but ayt i will do separate branches and pull request next time

**Pull request merged**